### ISSUE TITLE
Path is not taking as given and saving in /Download folder

### ISSUE THREAD

**Issue by pratik7span** [fluttercommunity/flutter_downloader#533]
I am getting issues while downloading videos from the server. I had given the path for Android as (`/storage/emulated/0/Movies/MyApp`) but It's downloading video in the "Download" folder only.

Here is my code:

```
requestDownloadFile(String link, bool isShare) async {
    _permissionReady = await _checkPermission();

    if (_permissionReady) {
      FlutterDownloader.registerCallback(callback);

      var _directory = Platform.isAndroid ? Directory('/storage/emulated/0/Movies/MyApp') : await getApplicationDocumentsDirectory();

      String newPath = _directory.absolute.path;

      print("Absolute Path: $newPath");

      if(Platform.isIOS) {
        newPath += '/MyApp';
      }
      _directory = Directory(newPath);

      bool hasExisted = await _directory.exists();
      if (!hasExisted) {
        _directory.create();
      }

      _localPath = _directory.path;

      print(_localPath);

      String fileName = link.split('/').last;

      print(fileName);

      print(_localPath + '/' + fileName);

      String localFilePath = _localPath + '/' + fileName;

      if (await File(localFilePath).exists() && isShare) {
        Share.shareFiles([localFilePath]);
      } else {
        ReceivePort _port = ReceivePort();
        IsolateNameServer.registerPortWithName(_port.sendPort, 'downloader_send_port');

        streamSubscription = _port.listen((dynamic data) {
          DownloadTaskStatus status = data[1];
          if (status.value == 3) {
            print(fileName);
            print('Downloaded Path: ${_localPath + ' / ' + fileName}');
            streamSubscription.cancel();
            IsolateNameServer.removePortNameMapping('downloader_send_port');

            // Share only if clicked on share otherwise download
            if (isShare) Share.shareFiles([_localPath + '/' + fileName]);
          }
        });

        print('Path Before Downloading: $_localPath');
        await FlutterDownloader.enqueue(url: link, savedDir: _localPath, showNotification: true, openFileFromNotification: true, fileName: fileName);
      }
    }
  }
```

Here are the logs:

```
I/flutter (25574): Absolute Path: /storage/emulated/0/Movies/MyApp
I/flutter (25574): Directory: '/storage/emulated/0/Movies/MyApp'
I/flutter (25574): /storage/emulated/0/Movies/MyApp
I/flutter (25574): gbUG2Ag6LHA4UGoyG2agOeV6Uu03Thn8RVcsTu1Q.mp4
I/flutter (25574): /storage/emulated/0/Movies/MyApp/gbUG2Ag6LHA4UGoyG2agOeV6Uu03Thn8RVcsTu1Q.mp4
I/flutter (25574): Path Before Downloading: /storage/emulated/0/Movies/MyApp

D/DownloadWorker(25574): DownloadWorker{url=https://example.com/gbUG2Ag6LHA4UGoyG2agOeV6Uu03Thn8RVcsTu1Q.mp4,filename=gbUG2Ag6LHA4UGoyG2agOeV6Uu03Thn8RVcsTu1Q.mp4,savedDir=/storage/emulated/0/Movies/MyApp,header=,isResume=false
```

But after downloading its printing (from the library) as below:

```
D/DownloadWorker(25574): File downloaded (/storage/emulated/0/Download/gbUG2Ag6LHA4UGoyG2agOeV6Uu03Thn8RVcsTu1Q.mp4)
D/DownloadWorker(25574): Setting an intent to open the file /storage/emulated/0/Download/gbUG2Ag6LHA4UGoyG2agOeV6Uu03Thn8RVcsTu1Q.mp4
```

**Can you please tell me why it's downloading the video in the Download folder only? Why it's not taking the path I have given?**

**Comment by abeyshiferaw0** [comment]
i am also facing the same issue, i wanted to download a file to private storage by path provided by getApplicationSupportDirectory but the package keeps downloading to downloads folder in the external storage

**Comment by DpkPatil05** [comment]
Same issue here

**Comment by abeyshiferaw0** [comment]
i have checked the native android implementation, this isn't a bug don't know why but the path provided is ignored and is replaced by downloads media folder intentionally,    

    @RequiresApi(Build.VERSION_CODES.Q)
    private Uri addFileToDownloadsApi29(String filename) {
        Uri collection = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
        try {
            ContentValues values = new ContentValues();
            values.put(MediaStore.Downloads.DISPLAY_NAME, filename);
            ContentResolver contentResolver = getApplicationContext().getContentResolver();
            return contentResolver.insert(collection, values);
        } catch (Exception e) {
            e.printStackTrace();
            log("Create a file using MediaStore API failed ");
        }
        return null;
    }

**Comment by abeyshiferaw0** [comment]
For any one who wants to the downloaded files to private or local storage 

1.go to the android implementation and look for DownloadWorkder.java file 

2.look for this function addFileToDownloadsApi21 and replace its content with  this
      File newFile = new File(getApplicationContext().getFilesDir(), filename);
      return newFile;
               
3. also  comment  this 
//                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
//                    uriApi29 = addFileToDownloadsApi29(filename);
//                    if (isResume) {
//                        outputStream = context.getContentResolver().openOutputStream(uriApi29, "wa");
//                    } else {
//                        outputStream = context.getContentResolver().openOutputStream(uriApi29, "w");
//                    }
//                } else {
//                    fileApi21 = addFileToDownloadsApi21(filename);
//                    outputStream = new FileOutputStream(fileApi21, isResume);
//                }

and add this to bottom

 fileApi21 = addFileToDownloadsApi21(filename);
 outputStream = new FileOutputStream(fileApi21, isResume);



Also comment out a big chunk of code in  => if (status == DownloadStatus.COMPLETE) {}. condition

seems reckless but its a fix, you can file your files in the local storage

**Comment by DpkPatil05** [comment]
Hey there,

Thanks for the reply

I’ll try these things out

Also, I have not tested it yet for iOS device, If this same issue is
replicated there. Will let you know soon.


Regards,
Deepak Patil

**Comment by ReaganRealones** [comment]
Thank you @abeyshiferaw0 , works perfectly. Though I was wondering when this can be fixed in the master repo

**Comment by Pitchoon** [comment]
same problem, any one finds a good workaround?

**Comment by AbdulrhmanBazrto** [comment]
having same issue here, how you solve it? @abeyshiferaw0 please can you explain how to go to "go to the android implementation" ? I'm still having same problem I can't store the files into subfolder of downlad directory

**Comment by yelkamel** [comment]
Same issue here

**Comment by abeyshiferaw0** [comment]
> > Also comment out a big chunk of code in => if (status == DownloadStatus.COMPLETE) {}. condition
> 
> Does it really work for you?

yes it does work, if you only want the downloaded file to be saved on the local private storage only

**Comment by abeyshiferaw0** [comment]
> having same issue here, how you solve it? @abeyshiferaw0 please can you explain how to go to "go to the android implementation" ? I'm still having same problem I can't store the files into subfolder of downlad directory

if your using android studio double click shift and search for downloadWorker.java

**Comment by abeyshiferaw0** [comment]
> Same issue here, can someone do a fork with the fix ? ^^'

https://github.com/abeyshiferaw0/flutter_downloader.git

this is the fork that i used, but is only for downloading to **_internal storage only_**

just path your desired location starting from getApplicationSupportDirectory() and it will work on both android and ios

e.g
` 
/// _to get downloaded file save path_
  Future<String> getSaveDir() async {
    Directory directory = await getApplicationSupportDirectory();
    Directory saveDir = Directory("${directory.absolute.path}${Platform.pathSeparator}SUB_FOLDER");
    bool exists = await saveDir.exists();
    if (!exists) {
      await saveDir.create(recursive: true);
    }
    return saveDir.path;
  }

/// _start download with path from getSaveDir()_
   FlutterDownloader.enqueue(
            url: URL,
            savedDir: getSaveDir(),
            fileName: FILE_NAME,
            showNotification: true,
            openFileFromNotification: false,
      );

`

**Comment by swarup13** [comment]
@hnvn  you can ask for this permission to create the folder
 await requestPermission(Permission.manageExternalStorage);

**Comment by Amanullahgit** [comment]
> addFileToDownloadsApi21

where can i find this file ?

**Comment by jr24jr** [comment]
Had the same problem and fixed it.
```
flutter_downloader: 1.7.0
permission_handler: 8.1.5
external_path: 1.0.1
```
and replace it with:
```
flutter_downloader: 1.5.2
permission_handler: 8.1.5
external_path: 1.0.1
```

to be able to write you must have permissions: storage, manageExternalStorage, requestInstallPackages.
in AndroidManifest.xml
```
<uses-permission android:name="android.permission.INTERNET" />
<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
<uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />
```

**Comment by digambernegi** [comment]
I guess removing or  changing "SaveInPublicStorage" to false would save the file to your required directory.

final id = await FlutterDownloader.enqueue(
          url: url,
          savedDir: externalDir.path,
          showNotification: true,
          openFileFromNotification: true,
          saveInPublicStorage: false);

**Comment by Ali-Kheiri** [comment]
@abeyshiferaw0
Is changing download worker java file currently working ? (api v 33)
